src/middleware/streamingRouter.js

```diff
@@ -0,0 +1,269 @@
+
+              // This event is fired once the response is fully-received and ready for URL rewriting
+              if (statusEvents.finishResponseAsString) {
+                const returnedResponse = statusEvents.finishResponseAsString(responseBodyAsString)
+                if (returnedResponse !== undefined && returnedResponse) {
+                  responseBodyAsString = returnedResponse
+                }
+              }
+              response.body = responseBodyAsString
+            } else {
+              response.body.push(null)
+            }
+
+            response.timestampEnd = new Date()
+
+            if (statusEvents.finishResponse) {
+              statusEvents.finishResponse(response, size)
+            }
+
+            const charset = obtainCharset(routeRes.headers)
+            if (routeRes.headers['content-encoding'] === 'gzip') {
+              gunzip.on('end', () => {
+                const uncompressedBody = Buffer.concat(uncompressedBodyBufs)
+                response.body = uncompressedBody.toString(charset)
+                resolve(response)
+              })
+            } else if (routeRes.headers['content-encoding'] === 'deflate') {
+              inflate.on('end', () => {
+                const uncompressedBody = Buffer.concat(uncompressedBodyBufs)
+                response.body = uncompressedBody.toString(charset)
+                resolve(response)
+              })
+            } else {
+              resolve(response)
+            }
+          })
+
+        // If request socket closes the connection abnormally
+        routeRes.connection
+          .on('error', (err) => {
+            if (statusEvents.responseError) {
+              statusEvents.responseError(err)
+            }
+            logger.error(`Connection Error on socket: ${err}`)
+            reject(err)
+          })
+      })
+      .on('error', (err) => {
+        if (statusEvents.responseError) {
+          statusEvents.responseError(err)
+        }
+        logger.error(`Error streaming response upstream: ${err}`)
+        reject(err)
+      })
+      .on('clientError', (err) => {
+        if (statusEvents.clientError) {
+          statusEvents.clientError(err)
+        }
+        logger.error(`Client error streaming response upstream: ${err}`)
+        reject(err)
+      })
+
+    const timeout = (options.timeout != undefined) && (options.timeout) ? options.timeout : +config.router.timeout
+    routeReq.setTimeout(timeout, () => {
+      const err = new Error(`Request took longer than ${timeout}ms`)
+      routeReq.destroy(err)
+      if (statusEvents.timeoutError) {
+        statusEvents.timeoutError(timeout)
+      }
+      reject(err)
+    })
+
+    downstream
+      .on('data', (chunk) => {
+        if (options.requestBodyRequired) {
+          routeReq.write(chunk)
+        }
+        if (!startedRequest && statusEvents.startRequest) {
+          statusEvents.startRequest()
+          startedRequest = true
+        }
+      })
```

Comment:
Here we could again `.pipe()` downstream to `routeReq` we could probably start the request as soon as we start piping instead of on the first chunk and then checking this every time.